x-pack/plugin/data-streams/src/test/java/org/elasticsearch/xpack/datastreams/action/DeleteDataStreamTransportActionTests.java

```diff
@@ -0,0 +1,208 @@
+            )
+        );
+    }
+
+    private SnapshotsInProgress.Entry createEntry(String dataStreamName, String repo, boolean partial) {
+        return new SnapshotsInProgress.Entry(
+            new Snapshot(repo, new SnapshotId("", "")),
+            false,
+            partial,
+            SnapshotsInProgress.State.STARTED,
+            Collections.emptyList(),
+            List.of(dataStreamName),
+            0,
+            1,
+            ImmutableOpenMap.of(),
+            null,
+            null,
+            null
+        );
+    }
+
+    public void testDeleteNonexistentDataStream() {
+        final String dataStreamName = "my-data-stream";
+        String[] dataStreamNames = { "foo", "bar", "baz", "eggplant" };
+        ClusterState cs = getClusterStateWithDataStreams(
+            List.of(
+                new Tuple<>(dataStreamNames[0], randomIntBetween(1, 3)),
+                new Tuple<>(dataStreamNames[1], randomIntBetween(1, 3)),
+                new Tuple<>(dataStreamNames[2], randomIntBetween(1, 3)),
+                new Tuple<>(dataStreamNames[3], randomIntBetween(1, 3))
+            ),
+            List.of()
+        );
+        DeleteDataStreamAction.Request req = new DeleteDataStreamAction.Request(new String[] { dataStreamName });
+        ClusterState newState = DeleteDataStreamTransportAction.removeDataStream(getMetadataDeleteIndexService(), cs, req);
+        assertThat(newState.metadata().dataStreams().size(), equalTo(cs.metadata().dataStreams().size()));
+        assertThat(
+            newState.metadata().dataStreams().keySet(),
+            containsInAnyOrder(cs.metadata().dataStreams().keySet().toArray(Strings.EMPTY_ARRAY))
+        );
+    }
+
+    @SuppressWarnings("unchecked")
+    private static MetadataDeleteIndexService getMetadataDeleteIndexService() {
+        MetadataDeleteIndexService s = mock(MetadataDeleteIndexService.class);
+        when(s.deleteIndices(any(ClusterState.class), any(Set.class))).thenAnswer(mockInvocation -> {
+            ClusterState currentState = (ClusterState) mockInvocation.getArguments()[0];
+            Set<Index> indices = (Set<Index>) mockInvocation.getArguments()[1];
+
+            final Metadata.Builder b = Metadata.builder(currentState.metadata());
+            for (Index index : indices) {
+                b.remove(index.getName());
+            }
+
+            return ClusterState.builder(currentState).metadata(b.build()).build();
+        });
+
+        return s;
+    }
+
+    /**
+     * Constructs {@code ClusterState} with the specified data streams and indices.
+     *
+     * @param dataStreams The names of the data streams to create with their respective number of backing indices
+     * @param indexNames  The names of indices to create that do not back any data streams
+     */
+    public static ClusterState getClusterStateWithDataStreams(List<Tuple<String, Integer>> dataStreams, List<String> indexNames) {
```

Comment:
I suggest moving this method to DataStreamTestHelper since it's used in a lot of tests in the server module.

Reply:
Good point. This is why ci/1 failed. I've addressed this.